### ISSUE TITLE
BUG: dtype mismatch error in vg.retarded_local_clustering()

### ISSUE THREAD

**Issue by pucicu** [pik-copan/pyunicorn#163]
When calling `vg.retarded_local_clustering()` a dtype mismatch error occurs:

`ValueError: Buffer dtype mismatch, expected 'ADJ_t' but got 'short'`

(the same for `vg.advanced_local_clustering()`)

**Comment by lenas95** [comment]
Hi @pucicu ,

could you please elaborate on which data you used and how your code snippet looks like, so that we can reproduce it?

Greetings

**Comment by pucicu** [comment]
The analysis here https://zenodo.org/record/5578298 worked in the past, but now pyunicorn now throws the error
`ValueError: Buffer dtype mismatch, expected 'ADJ_t' but got ‘short’`

Simply use the data and analysis from this repository and let me know if it works for you.

Thanks and best wishes
Norbert

**Comment by fkuehlein** [comment]
Update:
`VisibilityGraph.advanced_local_clustering()` throws the same error.

Problem appears to be within cython modules

error message:

```
src/pyunicorn/timeseries/visibility_graph.py in retarded_local_clustering(self)
    254         norm = retarded_degree * (retarded_degree - 1) / 2.
    255 
--> 256         _retarded_local_clustering(N, A, norm, retarded_clustering)
    257         return retarded_clustering
    258 

src/pyunicorn/timeseries/_ext/numerics.pyx in pyunicorn.timeseries._ext.numerics._retarded_local_clustering()

ValueError: Buffer dtype mismatch, expected 'ADJ_t' but got 'short'
```

**Comment by fkuehlein** [comment]
__problem:__

`pyunicorn/timeseries/visibility_graph.py` in `self.retarded_local_clustering()`
- (line 256) error when calling `_retarded_local_clustering(N, A, norm, retarded_clustering)`: `Buffer dtype mismatch, expected 'ADJ_t' but got 'short'`
- same problem with `self.advanced_local_clustering()`

__traceback:__

`pyunicorn/timeseries/_ext/numerics.pyx` in `_retarded_local_clustering(N, A, norm, retarded_clustering)`
- (line 984) adjacency matrix `A` expected to be of dtype `ADJ_t`

`pyunicorn/core/_ext/types.pxd` and `/types.py`
- adjacency matrix dtype is defined as `ADJ_t ~ int8` (why not use boolean dtype?)

`pyunicorn/timeseries/visibility_graph.py` in `self.__init__()`
- (line 99) adjacency matrix `A` is calculated with help of `vg.visibility_relations()`. Adding `print(A.dtype)` below returns `int8`.
- (line 104) `A` is then passed to `InteractingNetworks.__init__(self, A)`, (creating an instance of InteractingNetworks class, which is an instance of Network class). Adding `print(self.adjacency.dtype)` below however returns `int16` (equivalent to `short` in C)

`pyunicorn/core/network.py` in `self.set_adjacency()`
- (line 482) dtype of adjacency matrix is set to `int16` (or `int32` if N is bigger), which might be the problem

__additionally to problem with dtype of adjacency matrix__ `A`

`pyunicorn/timeseries/visibility_graph.py` in `self.retarded_local_clustering()`
- (line 256) `_retarded_local_clustering(N, A, norm, retarded_clustering)`  expects dtype `FIELD_t` for `norm` and `retarded_clustering` params, but gets `double`

`pyunicorn/core/_ext/types.pxd` and `/types.py`
- dtype for fields defined as `FIELD_t ~ float32` 

same for `self.advanced_local_clustering()`  


__quick fixes/workarounds__

_alternative a_

`pyunicorn/timeseries/visibility_graph.py` in `self.retarded_local_clustering()`
- (line 256) instead of `_retarded_local_clustering(N, A, norm, retarded_clustering)` 
    do `_retarded_local_clustering(N, A.astype("int8"), norm.astype("float32"), retarded_clustering.astype("float32"))`
- (line 277) same for `self.advanced_local_clustering()`

_alternative b_

`pyunicorn/timeseries/_ext/numerics.pyx` in `_retarded_local_clustering()`
- (lines 984-986) change dtype `ADJ_t` to `short` and `FIELD_t` to `double`
- (lines 1009-1011) same for `_advanced_local_clustering()`

**Comment by ntfrgl** [comment]
Thank you for the detailed analysis, @fkuehlein! You have identified two oversights in 3dab5bf89d2e224fc319ddd64aeeecc480f27fba.

The easy thing first: The original types for the arguments `norm` and `retarded_clustering` were the deprecated `FLOATTYPE=np.float`, which should have been translated as `DFIELD=np.float64` instead of `FIELD=np.float32` in the Cython function signature. I have fixed that now for all such cases in 2b489da9e504dfbdb621848bb82388f510c046ca.

For the argument `A`, the appropriate solution is to downcast to `ADJ_t` at the call site, using the helper [`core._ext.types.to_cy()`](https://github.com/pik-copan/pyunicorn/blob/master/src/pyunicorn/core/_ext/types.py#L39) that exists for this purpose. I will leave that to you, and kindly ask you to add a test that triggers all possible scalar types for the adjacency matrix, in a similar fashion to [`test_generic.simple_instances()`](https://github.com/pik-copan/pyunicorn/blob/2f770aea02838aa7b9549aefc308e4957309daab/tests/test_generic.py#L41). Please look out for other places where a similar pattern might occur.

---

There are several layers of causation to this bug, which I document here in hopes that it will inform the resolution and mitigation of other similar bugs:

- The offending commit included a large number of type conversions in an attempt to make the Python/Cython interface more coherent. This involved avoiding officially deprecated type aliases, making explicit many implicit type casts, and unifying previously inconsistent type uses throughout the codebase via semantic type aliases.
- Because of a significant lack of unit tests, in many cases these changes could not be automatically tested. This requires a systematic effort to extend the code coverage of unit tests to at least trigger all methods published in tutorials and in papers by the core authors.
- Furthermore, these changes have so far not been carefully reviewed due to a lack of resources. @fkuehlein, given the code analysis you have already done for this issue, it would be wonderful if you could extend your review to the entire associated change set.
- As you explained, the dynamically changing numeric type of sparse adjacency matrices is the specific issue underlying this bug. As far as I can tell, this is a choice from early on in the library's development and is motivated by some operations on sparse matrices, e.g., row-wise or column-wise sums, which would otherwise result in overflow w.r.t. the scalar type holding the binary values of the sparse matrix. Of course, circumventing the overflow problem by upcasting the binary values is a very resource-inefficient and not a failsafe solution. Ideally, the adjacency should always be represented using the smallest possible type and potentially overflowing sparse matrix operations should be rewritten, or at least implemented using only local and temporary upcasts. But this would constitute a significant change to the core of the library and would require extensive testing, reinforcing the point above.
- More generally, this issue is a prime example for the general need to either test much more rigorously or to rewrite the library in a strongly typed way. The offending commit attempted to find a compromise with very limited resources, by forcing type casts to fail only at the Python/Cython interface when they don't conform to some "standard" behaviour that was encoded explicitly in the semantic type aliases, under the assumption that no dynamic type changes of significance will occur inside the Cython functions.

**Comment by fkuehlein** [comment]
Thanks for your help and the detailed additional info, @ntfrgl!

The original issue should be resolved as of [`1651a2b`](https://github.com/fkuehlein/pyunicorn/commit/1651a2b443d40131d6adfea8efe2821bdddb5916). For now, that commit remains on my fork and I will create a pull request as soon as I have cleared at least the rest of `VisibilityGraph` of similar possible occurrences. 

I have two questions remaining in order to proceed on this:

__1. Appropriate solution for these cases__

As I have understood, the minimal solution (which you have pursued so far) would be to generally pass arguments `arg` of a cython module `_ext.numerics._mod(arg)` with help of  `_ext.numerics._mod(to_cy(arg, ARGTYPE))` to make sure they are cast to the expected type `ARGTYPE` (as is done throughout [`core.Network`](https://github.com/pik-copan/pyunicorn/blob/master/src/pyunicorn/core/network.py)). 

All remaining variables that might be affected by this within `VisibilityGraph`, however, are already explicitly set to the required dtype on initialization. Namely:

- The empty placeholder for the adjacency matrix [`A`](https://github.com/pik-copan/pyunicorn/blame/master/src/pyunicorn/timeseries/visibility_graph.py#L129), which is correctly initialized to be of `dtype=MASK`. I guess I keep that as it is.

- [`VisibilityGraph.time_series`](https://github.com/pik-copan/pyunicorn/blame/master/src/pyunicorn/timeseries/visibility_graph.py#L81) and [`VisibilityGraph.timings`](https://github.com/pik-copan/pyunicorn/blame/master/src/pyunicorn/timeseries/visibility_graph.py#L90), which are initialized to be of `dtype=np.float32`. Keep that, or put `dtype=FIELD` there instead?


__2. Implementation of a respective test__

I'm not sure what you mean by _"trigger all possible scalar types for the adjacency matrix"_. `self.adjacency` is a property of the `Network` class. So maybe the idea is to minimally initialize all classes that inherit from `Network` and check what scalar types they return for `self.adjacency`. Does that make sense? Or else, could you elaborate on that?

**Comment by ntfrgl** [comment]
**Short-term answer**
1. **Solution:**
   - Apart from mistakes like the ones above, the original commit introducing semantic type aliases followed a manual, defensive approach. The goal was to use typed initialisation where it is apparently valid (i.e., the variable is initialised within the current scope and no type-altering methods are applied before the Cython call), and to type cast directly at the call site otherwise.
   - This means that most of the work should already be done, but it is expected that a few mistakes not triggered in the current test suite need to be fixed, and that some unnecessary copy/cast operations might be avoidable with a more precise type inspection.
   - It would be preferable to replace the remaining non-alias types with the semantic aliases where possible.
3. **Testing:**
   - In the current bug's case, the problematic type mutation occurs depending on the node count, which means that in order to guard against future instances of similar bugs, the test for this method should be parametrised with several different network sizes to ensure that all possible cases of Cython calls are covered.
   - The reliability and reusability of the test suite would be improved if this was done using [parametric test fixtures](https://docs.pytest.org/en/stable/example/parametrize.html), which exist for such scenarios.
   - And btw, if in the process of looking for other similar situations you find it helpful to search for `ADJ_t`, then the semantic type aliases are fulfilling part of their purpose.

**Long-term answer** *(worth discussing and creating a new issue for)*
1. **Solution:**
   - The above is really an incomplete substitute for a proper type analysis, which has traditionally not been available in the dynamically typed Python language, and is held together by non-mechanised reasoning and incomplete testing. Given recent developments in the Python language and its library ecosystem, a much more reliable and efficient approach would be to use the new [NumPy API type annotations](https://numpy.org/devdocs/reference/typing.html). These could be introduced incrementally throughout the codebase, and would allow static type checking of the numerical types. I expect that this will be straightforward for the most part with few exceptions, that it might expose some well hidden bugs, and that it would make the library a lot more sustainable.
3. **Testing:**
   - In addition to the above extensions of the unit tests, [Mypy](https://mypy-lang.org/) would execute the static type checking in combination with the [plugin](https://numpy.org/devdocs/reference/typing.html#mypy-plugin) for NumPy API types.

**Comment by fkuehlein** [comment]
Ok, actually found another `dtype=np.uint8` in [`timeseries._ext.numerics._visibility()`](https://github.com/pik-copan/pyunicorn/blob/master/src/pyunicorn/timeseries/_ext/numerics.pyx#L981). The function in question is only used within [`test_timeseries.testVisibility()`](https://github.com/pik-copan/pyunicorn/blob/master/tests/test_timeseries/test_timeseries.py#L185) though, which is currently commented out. 

Fixed both the test and the cython function in question with [`bb5e2ac`](https://github.com/fkuehlein/pyunicorn/commit/bb5e2acbfc68fe19ed1b0d5370e24fbecaf54325) for a start, on a new branch. Still, the test only covers a specific testing method, so test coverage is lacking (let alone parametrisation). I will therefore further look into enhancements.

Putting this here just for documentation.

**Comment by ntfrgl** [comment]
Let me add to the critical observations in your latest commit message, since it doesn't have an associated issue yet:
- `_visibility()`: The core of this function is a loop with a 1-bit accumulator, and should not be allocating an array and calling several NumPy methods on it.
- `testVisibility()`: A test which essentially replicates the code under test is almost useless.

**Comment by ntfrgl** [comment]
Resolved by #183.
